Publish public preview swagger spec for "Azure metrics Ingestion REST  API"

vickyv51 opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [Done] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

vickyv51 committed:
-Publish "Azure metrics REST Ingestion API" public preview swagger spec

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3856) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3856) before we can accept your contribution.<br/><br/>:x: vickyv51 [sign now](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3856)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-rest-api-specs?pullRequest=3856) it.</sub>

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-python

azuresdkci commented:
Can one of the admins verify this patch?

jhendrixMSFT reviewed (changes requested):
Missing README.md file, see [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/monitor/resource-manager/readme.md) for an example.

jhendrixMSFT commented on specification/monitor/data-plane/preview/2018-09-01-preview.json:
```diff
@@ -0,0 +1,1927 @@
+              "$ref": "#/definitions/Microsoft.AspNetCore.Mvc.ObjectResult"
+            }
+          },
+          "400": {
```
Shouldn't need to model these failure cases, use ```default``` instead.  See [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/datalake-store/data-plane/Microsoft.DataLakeStore/stable/2016-11-01/filesystem.json#L95) for an example.
Also is it expected that the 200 and 4xx cases return the same model?

jhendrixMSFT replied:
Ideally this information would be returned in the response body as an [OData v4 object](http://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091).  Something like this.
```json
{
  "Error": {
    "Code": "InvalidRequest",
    "Message": "Some descriptive error message."
  }
}
```

vickyv51 replied:
We indeed already do that. So is the suggestion that having above behavior good enough?

jhendrixMSFT replied:
I would suggest then that you specify a ```default``` response that references the appropriate model definition.  Looking at the definition for ```Microsoft.AspNetCore.Mvc.ObjectResult``` it's not an OData v4 error object (or a superset of it).

jhendrixMSFT replied:
More importantly, the code generators will treat any status code modeled here as a success.  So if these 4xx codes aren't to be treated as success then you will want to remove them.

vickyv51 replied:
Makes sense. Fixing it. :)

jhendrixMSFT commented on specification/monitor/data-plane/preview/2018-09-01-preview.json:
```diff
@@ -0,0 +1,1927 @@
+    "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{resourceProvider}/{resourceTypeName}/{resourceName}/metrics": {
+      "post": {
+        "summary": "Ingestion API used by Azure custom metrics\r\n            Responsible of below operation\r\n            1. Authorize the request\r\n            2. Deserialize payload request body\r\n            3. Perform validation of the payload\r\n            4. Commits the payload for metrics ingestion",
+        "operationId": "postSubscriptionsBySubscriptionIdResourcegroupsByResourceGroupNameProvidersByResourceProviderByResourceTypeNameByResourceNameMetrics",
```
The ```operationId``` is used to generate method names. Please shorten it to something more descriptive.

jhendrixMSFT commented on specification/monitor/data-plane/preview/2018-09-01-preview.json:
```diff
@@ -0,0 +1,1927 @@
+    "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{resourceProvider}/{resourceTypeName}/{resourceName}/metrics": {
+      "post": {
+        "summary": "Ingestion API used by Azure custom metrics\r\n            Responsible of below operation\r\n            1. Authorize the request\r\n            2. Deserialize payload request body\r\n            3. Perform validation of the payload\r\n            4. Commits the payload for metrics ingestion",
+        "operationId": "postSubscriptionsBySubscriptionIdResourcegroupsByResourceGroupNameProvidersByResourceProviderByResourceTypeNameByResourceNameMetrics",
```
Please provide an example for this operation, see [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/datalake-store/data-plane/Microsoft.DataLakeStore/stable/2016-11-01/filesystem.json#L102) for an example.

vickyv51 replied:
Updating in next iteration

vickyv51 committed:
Address review comments
 - Use a shorter operationId
 - Remove any non-success response code with default response clause
 - Add a new example and reference it in the swagger spec
 - Include readme.md file for the swagger spec

vickyv51 committed:
Fix a bug in readme.md.
 - Inline the readme.python.md file in the readme.md

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-java

vickyv51 commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,157 @@
+output-folder: $(azure-libraries-for-java-folder)/azure-mgmt-monitor
+```
+
+# Validation
````
Thats correct. Dropping it.

vickyv51 committed:
-Remove validation section that is not necessary

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+
+``` yaml $(tag) == 'package-2018-09'
+input-file:
+- data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json
````
This path is relative to the location of the README.md file, it should be `preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json`.

vickyv51 replied:
My bad. Thanks for catching it. Fixed

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+```
+### Tag: package-2018-09
+
+These settings apply only when `--tag=package-2018-09` is specified on the command line.
````
I would add a `-preview` suffix to the tag name so it's clear this is preview, plus if you end up releasing a 2018-09-01 GA the tag names will be unique.

vickyv51 replied:
Makes sense. Fixing it.

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
```diff
@@ -0,0 +1,147 @@
+  azure-arm: true
+  payload-flattening-threshold: 1
+  license-header: MICROSOFT_MIT_NO_VERSION
+  namespace: Microsoft.Azure.Management.Monitor
```
Is this the correct namespace for the C# implementation?  @shahabhijeet can you review this part?

jhendrixMSFT replied:
Given that this is data-plane I would think the word `Management` doesn't belong in the namespace but I'm not really sure.

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+output-folder: $(go-sdk-folder)/services/preview/monitor/mgmt/2018-03-01/insights
+```
+
+## Python
````
@lmazuel can you please review the config section for python?

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+  output-folder: $(python-sdks-folder)/azure-mgmt-monitor
+```
+
+## Java
````
@jianghaolu can you please review the config section for Java?

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+``` yaml
+title: MonitorClient
+description: Monitor Management Client
+openapi-type: arm
````
The value here should be `data-plane` not `arm`.

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+``` yaml $(tag) == 'package-2018-09' && $(go)
+output-folder: $(go-sdk-folder)/services/preview/monitor/mgmt/2018-03-01/insights
````
Value is wrong, this should be `$(go-sdk-folder)/services/preview/monitor/2018-09-01-preview/monitor`

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+``` yaml $(go)
+go:
+  license-header: MICROSOFT_APACHE_NO_VERSION
+  namespace: insights
````
If the package directory is going to be `monitor` then this namespace will need to match it.

vickyv51 committed:
-Address review comments & fix several copy paste mistakes to use sensible names instead

jhendrixMSFT commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+``` yaml $(go)
+go:
+  license-header: MICROSOFT_APACHE_NO_VERSION
+  namespace: azuremonitor
````
This needs to match the package directory (i.e. the leaf directory on line 104).

lmazuel commented on specification/monitor/data-plane/readme.md:
```diff
@@ -0,0 +1,147 @@
+  azure-arm: true
+  license-header: MICROSOFT_MIT_NO_VERSION
+  payload-flattening-threshold: 2
+  namespace: microsoft.azure.monitor.metrics
```
azure.monitor

lmazuel commented on specification/monitor/data-plane/readme.md:
```diff
@@ -0,0 +1,147 @@
+  license-header: MICROSOFT_MIT_NO_VERSION
+  payload-flattening-threshold: 2
+  namespace: microsoft.azure.monitor.metrics
+  package-name: microsoft.azure.monitor.metrics
```
azure-monitor

vickyv51 replied:
I'm fixing this assuming below
namespace: azure.monitor
package-name: azure-monitor

lmazuel commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+``` yaml $(python) && $(python-mode) == 'update'
+python:
+  no-namespace-folders: true
+  output-folder: $(python-sdks-folder)/azure-monitor/azure/dataplane/monitor
````
output-folder: $(python-sdks-folder)/azure-monitor/azure/monitor

lmazuel commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+``` yaml $(python) && $(python-mode) == 'create'
+python:
+  basic-setup-py: true
+  output-folder: $(python-sdks-folder)/azure-monitor/azure/dataplane/monitor
````
output-folder: $(python-sdks-folder)/azure-monitor/

jhendrixMSFT commented on specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json:
```diff
@@ -0,0 +1,1896 @@
+            }
+          }
+        },
+        "x-ms-example": {
```
Should be `x-ms-examples`

vickyv51 commented on specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json:
```diff
@@ -0,0 +1,1896 @@
+    }
+  },
+  "definitions": {
+    "Microsoft.Azure.Monitoring.Ingestion.Gateway.Models.AzureMetricsDocument": {
```
These are actually generated directly using swagger extension based on the code using VS annotations. I will drop the prefix to make it more non c#-ish

vickyv51 committed:
Address the PR comments
  - Fix the readme.md naming incompatibilities
  - Remove the c#-ish namespace prefix

jhendrixMSFT commented on specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json:
```diff
@@ -0,0 +1,1896 @@
+        }
+      }
+    },
+    "Microsoft.AspNetCore.Mvc.ObjectResult": {
```
More prefix to remove, also on line 208.

jhendrixMSFT commented:
Please take a look at the [linter](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/427443386) errors and [model validation](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/427443392) failures.

vickyv51 commented on specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json:
```diff
@@ -0,0 +1,1896 @@
+        }
+      }
+    },
+    "Microsoft.AspNetCore.Mvc.Formatters.IOutputFormatter": {
```
Fixed.

vickyv51 committed:
-Remove assembly namespace prefix in few other places
-Rename bunch of properties to follow camel case naming
-Update the enum types to have the "x-ms-enum" part of the json

jianghaolu commented on specification/monitor/data-plane/readme.md:
````diff
@@ -0,0 +1,147 @@
+``` yaml $(java)
+azure-arm: true
+fluent: true
+namespace: com.microsoft.azure.dataplane.monitor
````
simply "com.microsoft.azure.monitor"

jianghaolu commented on specification/monitor/data-plane/readme.md:
```diff
@@ -0,0 +1,147 @@
+namespace: com.microsoft.azure.dataplane.monitor
+license-header: MICROSOFT_MIT_NO_CODEGEN
+payload-flattening-threshold: 1
+output-folder: $(azure-libraries-for-java-folder)/azure-monitor
```
$(azure-libraries-for-java-folder)/monitor/data-plane

vickyv51 committed:
-Fix last batch of ARM Errors
-Update path and namespace for Java section

jhendrixMSFT commented:
There are multiple enum types named `attributes` that have different sets of values which is not allowed, see [this](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/427946630) build log.  Either they all need to have the same set of values (which in that case just define it once and reference it) or their names need to be different.
Also you can run some of these validation tools locally, see https://github.com/Azure/oav for details.

vickyv51 committed:
-Drop all the types that recursively came with ObjectResult and use simple object type instead. This should make the swagger spec clean and more easily maintainable in the long run

jhendrixMSFT commented:
Down to just two errors in the model validation regarding the example, see the [log](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/428294649).

vickyv51 committed:
-Fix the example to just follow key-value pairs for all required parameters.

vickyv51 committed:
-Fix the responses section in the example

vickyv51 committed:
-Fix the type bug in the example, use int32 instead of string
